app/src/main/java/com/example/hackernews/viewmodel/MainViewModel.kt

```diff
@@ -48,12 +66,38 @@ class MainViewModel(private val newsRepository: NewsRepository) : ViewModel() {
                 handleError(error)
                 return@getNews
             }
-
             _topStories.add(news!!)
-            allNews.value = _topStories
+            this.news.value = _topStories
         }
     }
 
+    fun loadSavedStories() {
+        userRepository.loadSavedStories() { userSavedNews ->
+            _savedNews.postValue(userSavedNews)
+        }
+    }
+
+    fun saveStory(currentElement: Int) {
+        if (userRepository.currentUser() != null) {
+            userRepository.saveStory(
+                userRepository.currentUser()!!.user_id,
+                _topStories[currentElement].id
+            )
+            _hasNewsSaved.value = true
+            return
+        }
+        _hasNewsSaved.value = false
+    }
+
+    // providing these function to avoid Law of Demeter - don't talk to strangers!
```

Comment:
Good for reading up on this. But Law of Demeter states that component should not know the inner working of objects it manipulates so that would be applicable if you had a News object in ViewModel and tried to access its properties from view like viewmodel.news.newStoryId or something. 

not applicable in this case.